table/tables/partition.go

```diff
@@ -180,29 +223,48 @@ func partitionRecordKey(pid int64, handle int64) kv.Key {
 func (t *partitionedTable) locatePartition(ctx sessionctx.Context, pi *model.PartitionInfo, r []types.Datum) (int64, error) {
 	var err error
 	var isNull bool
-	partitionExprs := t.partitionExpr.UpperBounds
-	idx := sort.Search(len(partitionExprs), func(i int) bool {
+	idx := 0
+	switch t.meta.Partition.Type {
+	case model.PartitionTypeRange:
+		partitionExprs := t.partitionExpr.UpperBounds
+		idx = sort.Search(len(partitionExprs), func(i int) bool {
+			var ret int64
+			ret, isNull, err = partitionExprs[i].EvalInt(ctx, chunk.MutRowFromDatums(r).ToRow())
+			if err != nil {
+				return true // Break the search.
+			}
+			if isNull {
+				// If the column value used to determine the partition is NULL, the row is inserted into the lowest partition.
+				// See https://dev.mysql.com/doc/mysql-partitioning-excerpt/5.7/en/partitioning-handling-nulls.html
+				return true // Break the search.
+			}
+			return ret > 0
+		})
+		if err != nil {
+			return 0, errors.Trace(err)
+		}
+		if isNull {
+			idx = 0
+		}
+		if idx < 0 || idx >= len(partitionExprs) {
+			// The data does not belong to any of the partition?
+			return 0, errors.Trace(table.ErrTrgInvalidCreationCtx)
+		}
+
+	case model.PartitionTypeHash:
```

Comment:
MySQL also supports linear hashing, do we need support? @tiancaiamao

Reply:
`LINEAR HASH Partitioning `

```
CREATE TABLE employees (
    id INT NOT NULL,
    fname VARCHAR(30),
    lname VARCHAR(30),
    hired DATE NOT NULL DEFAULT '1970-01-01',
    separated DATE NOT NULL DEFAULT '9999-12-31',
    job_code INT,
    store_id INT
)
PARTITION BY LINEAR HASH( YEAR(hired) )
PARTITIONS 4;
```

Reply:
`LINEAR HASH Partitioning `

```
CREATE TABLE employees (
    id INT NOT NULL,
    fname VARCHAR(30),
    lname VARCHAR(30),
    hired DATE NOT NULL DEFAULT '1970-01-01',
    separated DATE NOT NULL DEFAULT '9999-12-31',
    job_code INT,
    store_id INT
)
PARTITION BY LINEAR HASH( YEAR(hired) )
PARTITIONS 4;
```

Reply:
Support LINEAR HASH is easy, just some small difference between the  hash function.